### PR TITLE
dialects: add arith canoncialization pattern "AdditionOfSameVariablesToMultiplybyTwo"

### PR DESCRIPTION
**Pull request by dshaaban01**
This arith rewrite pattern does the following:

`a + a -> a * 2`

Update: @math-fehr says we may not want to merge "bad" rewrite patterns. Please see [this PR](https://github.com/xdslproject/xdsl/pull/2167) as a potential solution.

### CODE REVIEW

File: tests/filecheck/transforms/arith-addition-of-same-variables-to-multiply-by-two.mlir [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// RUN: xdsl-opt %s -p canonicalize | filecheck %s
+
+func.func @hello(%n : i32) -> i32 {
+  %two = arith.constant 3 : i32
```
Comment (superlopuh) [inline tests/filecheck/transforms/arith-addition-of-same-variables-to-multiply-by-two.mlir:4]: 🤔

File: tests/filecheck/transforms/arith-addition-of-same-variables-to-multiply-by-two.mlir
Comment (superlopuh) [inline tests/filecheck/transforms/arith-addition-of-same-variables-to-multiply-by-two.mlir:1]: Can you please put all canonicalization tests in the same file, like for the riscv canonicalization?

**Comment by webmiche** [comment]
Why not `a + a -> a << 1`? I guess a shift would even be faster than an add usually.

**Comment by superlopuh** [comment]
Is this related to the other PR? I would also prefer shifting, but overall it feels like this may be better in an xdsl-gui-specific file somewhere, unless MLIR also has the same canonicalization. The last bit should be easy to test, just run `mlir-opt --canonicalize` on the test file.

**Comment by webmiche** [comment]
> Is this related to the other PR? 

My comment? Not really, just bikeshedding ideas. Are we reimplementing mlir canonicalizations on arith here? Or general ones? I am confused :sweat_smile: 

I guess we might only want to do the same canonicalizations as MLIR, out-of-the box at least.

**Comment by superlopuh** [comment]
No I meant to direct the question to Dalia

**Comment by superlopuh** [comment]
I think we're on the same page, I wouldn't be comfortable adding a canonicalization that's not in MLIR

**Pull request closed without merging**